File: sdk/python/inkbox/slack.py
"""Live Slack connections and messaging. Use invitation URLs for browser onboarding."""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Literal, TypedDict
from urllib.parse import quote
from uuid import UUID

from inkbox._http import HttpTransport

SlackMessageKind = Literal["dm", "group_dm", "mention", "channel", "thread"]


class SlackWebhookFilter(TypedDict, total=False):
    """Selectors combine with AND; message kinds combine with OR. Null means all."""

    connection_ids: list[UUID | str] | None
    conversation_ids: list[str] | None
    message_kinds: list[SlackMessageKind] | None


@dataclass
class SlackConnection:
    id: UUID
    identity_id: UUID
    workspace_id: str
    workspace_name: str
    bot_user_id: str
    status: Literal["connected", "disconnected", "reauthorization_required"]
    scopes: list[str]
    created_at: datetime


@dataclass
class SlackConnectionsResponse:
    connections: list[SlackConnection]
    installation_available: bool


@dataclass
class SlackInvitation:
    id: UUID
    identity_id: UUID
    status: str
    expires_at: datetime
    invitation_url: str | None = None


@dataclass
class SlackAction:
    id: UUID
    connection_id: UUID
    status: Literal["sending", "sent", "failed", "unknown"]
    conversation_id: str
    message_ts: str | None = None
    thread_ts: str | None = None
    error_code: str | None = None


@dataclass
class SlackConversationsResponse:
    conversations: list[dict[str, Any]]
    next_cursor: str | None = None


@dataclass
class SlackMessagesResponse:
    messages: list[dict[str, Any]]
    next_cursor: str | None = None
    has_more: bool = False


@dataclass
class SlackFile:
    id: str
    name: str | None = None
    title: str | None = None
    mimetype: str | None = None
    size: int | None = None
    downloadable: bool = False


def _parse(cls, raw):
    data = {k: v for k, v in raw.items() if k in cls.__dataclass_fields__}
    for key in ("id", "identity_id", "connection_id"):
        if key in data and cls is not SlackFile:
            data[key] = UUID(data[key])
    for key in ("created_at", "expires_at"):
        if key in data:
            data[key] = datetime.fromisoformat(data[key])
    return cls(**data)


def _connection(connection_id: UUID | str) -> str:
    return f"/slack/connections/{quote(str(connection_id), safe='')}"


class SlackResource:
    """No eager pagination or automatic resend of ambiguous send outcomes."""

    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def list_connections(self, identity_id: UUID | str) -> SlackConnectionsResponse:
        data = self._http.get(
            "/slack/connections", params={"identity_id": str(identity_id)}
        )
        return SlackConnectionsResponse(
            [_parse(SlackConnection, r) for r in data["connections"]],
            data["installation_available"],
        )

    def create_invitation(
        self, identity_id: UUID | str, *, expires_in_seconds: int = 86400
    ) -> SlackInvitation:
        """Create a one-time invitation URL to open in a browser (organization management only).

        Direct installation/accept requests require a cookie in that browser, so
        these browser-only endpoints are intentionally not exposed here.
        """
        return _parse(
            SlackInvitation,
            self._http.post(
                "/slack/invitations",
                json={
                    "identity_id": str(identity_id),
                    "expires_in_seconds": expires_in_seconds,
                },
            ),
        )

    def list_invitations(self, identity_id: UUID | str) -> list[SlackInvitation]:
        return [
            _parse(SlackInvitation, r)
            for r in self._http.get(
                "/slack/invitations", params={"identity_id": str(identity_id)}
            )
        ]

    def revoke_invitation(self, invitation_id: UUID | str) -> SlackInvitation:
        return _parse(
            SlackInvitation,
            self._http.post(
                f"/slack/invitations/{quote(str(invitation_id), safe='')}/revoke"
            ),
        )

    def disconnect(self, connection_id: UUID | str) -> SlackConnection:
        """Remove Inkbox authority locally; this does not uninstall the Slack app."""
        return _parse(
            SlackConnection, self._http.post(f"{_connection(connection_id)}/disconnect")
        )

    def list_conversations(
        self, connection_id: UUID | str, *, limit: int = 100, cursor: str | None = None
    ) -> SlackConversationsResponse:
        return _parse(
            SlackConversationsResponse,
            self._http.get(
                f"{_connection(connection_id)}/conversations",
                params={"limit": limit, "cursor": cursor},
            ),
        )

    def open_conversation(
        self, connection_id: UUID | str, user_ids: list[str]
    ) -> dict[str, Any]:
        return self._http.post(
            f"{_connection(connection_id)}/conversations", json={"user_ids": user_ids}
        )

    def get_conversation(
        self, connection_id: UUID | str, conversation_id: str
    ) -> dict[str, Any]:
        return self._http.get(
            f"{_connection(connection_id)}/conversations/{quote(conversation_id, safe='')}"
        )

    def list_messages(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        *,
        limit: int = 15,
        cursor: str | None = None,
        thread_ts: str | None = None,
    ) -> SlackMessagesResponse:
        if thread_ts is not None and not isinstance(thread_ts, str):
            raise ValueError("thread_ts must be a string")
        return _parse(
            SlackMessagesResponse,
            self._http.get(
                f"{_connection(connection_id)}/conversations/{quote(conversation_id, safe='')}/messages",
                params={"limit": limit, "cursor": cursor, "thread_ts": thread_ts},
            ),
        )

    def send_message(
        self,
        connection_id: UUID | str,
        *,
        conversation_id: str,
        text: str,
        idempotency_key: str,
        thread_ts: str | None = None,
    ) -> SlackAction:
        """Use a stable key for this exact message; poll get_action while sending; unknown is terminal uncertainty, never blindly resend."""
        if not isinstance(idempotency_key, str) or not re.fullmatch(
            r"[A-Za-z0-9._:-]{1,128}", idempotency_key
        ):
            raise ValueError(
                "idempotency_key must be 1..128 letters, digits, '.', '_', ':', or '-'"
            )
        body = {"conversation_id": conversation_id, "text": text}
        if thread_ts is not None:
            if not isinstance(thread_ts, str):
                raise ValueError("thread_ts must be a string")
            body["thread_ts"] = thread_ts
        return _parse(
            SlackAction,
            self._http.post(
                f"{_connection(connection_id)}/messages",
                json=body,
                headers={"Idempotency-Key": idempotency_key},
            ),
        )

    def get_action(
        self, connection_id: UUID | str, action_id: UUID | str
    ) -> SlackAction:
        return _parse(
            SlackAction,
            self._http.get(
                f"{_connection(connection_id)}/actions/{quote(str(action_id), safe='')}"
            ),
        )

    def get_file(self, connection_id: UUID | str, file_id: str) -> SlackFile:
        return _parse(
            SlackFile,
            self._http.get(
                f"{_connection(connection_id)}/files/{quote(file_id, safe='')}"
            ),
        )

    def download_file(self, connection_id: UUID | str, file_id: str) -> bytes:
        return self._http.get_bytes(
            f"{_connection(connection_id)}/files/{quote(file_id, safe='')}/content",
            accept="application/octet-stream",
        )

File: cli/src/commands/slack.ts
import { registerSlackOperationCommands } from "./slack-operations.js";
import { writeFile } from "node:fs/promises";
import { Command, InvalidArgumentError } from "commander";
import { createClient, getGlobalOpts } from "../client.js";
import { output } from "../output.js";
import { withErrorHandler } from "../errors.js";

export function slackInteger(value: string): number {
  if (
    !/^[0-9]+$/.test(value) ||
    !Number.isSafeInteger(Number(value)) ||
    Number(value) < 1
  )
    throw new InvalidArgumentError("Expected a positive integer");
  return Number(value);
}
interface IdentityOptions {
  identity?: string;
  identityId?: string;
}
const identity = (c: Command): Command =>
  c
    .option("-i, --identity <handle>", "Owning Inkbox identity handle")
    .option("--identity-id <id>", "Owning identity UUID instead of --identity");
async function resolveIdentityId(
  client: ReturnType<typeof createClient>,
  options: IdentityOptions,
): Promise<string> {
  if (!!options.identity === !!options.identityId)
    throw new InvalidArgumentError(
      "Provide exactly one of --identity or --identity-id",
    );
  if (options.identityId) return options.identityId;
  return (await client.getIdentity(options.identity!)).id;
}
const connection = (c: Command): Command =>
  c.requiredOption("--connection-id <id>", "Slack workspace connection UUID");
const conversation = (c: Command): Command =>
  connection(c).requiredOption(
    "--conversation-id <id>",
    "Slack conversation ID",
  );
const page = (c: Command): Command =>
  c
    .option("--limit <n>", "One page size", slackInteger)
    .option("--cursor <cursor>", "Next-page cursor; no automatic pagination");

export function registerSlackCommands(program: Command): void {
  const slack = program
    .command("slack")
    .description(
      "Slack workspace connections, live conversations, messages, and files",
    );
  const connections = slack
    .command("connection")
    .description("Manage workspace connections");
  identity(
    connections
      .command("list")
      .description("List connections and installation availability"),
  ).action(
    withErrorHandler(async function (this: Command, o: IdentityOptions) {
      const opts = getGlobalOpts(this);
      const client = createClient(opts);
      output(
        await client.slack.listConnections(await resolveIdentityId(client, o)),
        {
          json: !!opts.json,
        },
      );
    }),
  );
  connection(
    connections
      .command("disconnect")
      .description("Remove Inkbox authority; does not uninstall the Slack app"),
  ).action(
    withErrorHandler(async function (
      this: Command,
      o: { connectionId: string },
    ) {
      const opts = getGlobalOpts(this);
      output(await createClient(opts).slack.disconnect(o.connectionId), {
        json: !!opts.json,
      });
    }),
  );
  const invites = slack
    .command("invitation")
    .description("Organization management: workspace installation invitations");
  identity(
    invites
      .command("create")
      .description(
        "Create a one-time invitation URL; open it in the installer's browser",
      ),
  )
    .option(
      "--expires-in-seconds <n>",
      "Invitation lifetime: 300..604800 seconds (default 86400)",
      slackInteger,
    )
    .action(
      withErrorHandler(async function (
        this: Command,
        o: IdentityOptions & { expiresInSeconds?: number },
      ) {
        const opts = getGlobalOpts(this);
        const client = createClient(opts);
        output(
          await client.slack.createInvitation(
            await resolveIdentityId(client, o),
            o,
          ),
          { json: !!opts.json },
        );
      }),
    );
  identity(
    invites
      .command("list")
      .description(
        "List invitation status; links are only returned at creation",
      ),
  ).action(
    withErrorHandler(async function (this: Command, o: IdentityOptions) {
      const opts = getGlobalOpts(this);
      const client = createClient(opts);
      output(
        await client.slack.listInvitations(await resolveIdentityId(client, o)),
        {
          json: !!opts.json,
        },
      );
    }),
  );
  const installations = slack
    .command("installation")
    .description("Organization management: browser installation handoff");
  identity(
    installations
      .command("start")
      .description(
        "Return a short-lived secret URL to open in a browser; do not log or share it",
      ),
  )
    .option("--workspace-id <id>", "Expected Slack workspace ID")
    .action(
      withErrorHandler(async function (
        this: Command,
        o: IdentityOptions & { workspaceId?: string },
      ) {
        const opts = getGlobalOpts(this);
        const client = createClient(opts);
        output(
          await client.slack.startInstallation(
            await resolveIdentityId(client, o),
            o,
          ),
          { json: !!opts.json },
        );
      }),
    );
  invites
    .command("revoke <invitation-id>")
    .description("Revoke an unconsumed invitation")
    .action(
      withErrorHandler(async function (this: Command, id: string) {
        const opts = getGlobalOpts(this);
        output(await createClient(opts).slack.revokeInvitation(id), {
          json: !!opts.json,
        });
      }),
    );
  const conversations = slack
    .command("conversation")
    .description("Read live conversations and open direct messages");
  page(connection(conversations.command("list"))).action(
    withErrorHandler(async function (
      this: Command,
      o: { connectionId: string; limit?: number; cursor?: string },
    ) {
      const opts = getGlobalOpts(this);
      output(
        await createClient(opts).slack.listConversations(o.connectionId, o),
        { json: !!opts.json },
      );
    }),
  );
  conversation(conversations.command("get")).action(
    withErrorHandler(async function (
      this: Command,
      o: { connectionId: string; conversationId: string },
    ) {
      const opts = getGlobalOpts(this);
      output(
        await createClient(opts).slack.getConversation(
          o.connectionId,
          o.conversationId,
        ),
        { json: !!opts.json },
      );
    }),
  );
  connection(
    conversations
      .command("open")
      .description("Open a DM or group DM with 1..8 Slack users"),
  )
    .requiredOption(
      "--user-id <id>",
      "Slack user ID (repeatable)",
      (v: string, previous: string[] = []) => [...previous, v],
    )
    .action(
      withErrorHandler(async function (
        this: Command,
        o: { connectionId: string; userId: string[] },
      ) {
        const opts = getGlobalOpts(this);
        output(
          await createClient(opts).slack.openConversation(
            o.connectionId,
            o.userId,
          ),
          { json: !!opts.json },
        );
      }),
    );
  const messages = slack
    .command("message")
    .description(
      "Read live history and send messages with stable idempotency keys",
    );
  page(conversation(messages.command("list")))
    .option("--thread-ts <timestamp>", "Slack thread timestamp (string)")
    .action(
      withErrorHandler(async function (
        this: Command,
        o: {
          connectionId: string;
          conversationId: string;
          limit?: number;
          cursor?: string;
          threadTs?: string;
        },
      ) {
        const opts = getGlobalOpts(this);
        output(
          await createClient(opts).slack.listMessages(
            o.connectionId,
            o.conversationId,
            o,
          ),
          { json: !!opts.json },
        );
      }),
    );
  conversation(
    messages
      .command("send")
      .description(
        "Send once; unknown is unresolved, not safe to blindly resend",
      ),
  )
    .requiredOption("--text <text>", "Message text (1..12000 characters)")
    .requiredOption(
      "--idempotency-key <key>",
      "Stable key for this exact operation; reuse only with the same body",
    )
    .option("--thread-ts <timestamp>", "Reply thread timestamp (string)")
    .action(
      withErrorHandler(async function (
        this: Command,
        o: {
          connectionId: string;
          conversationId: string;
          text: string;
          idempotencyKey: string;
          threadTs?: string;
        },
      ) {
        const opts = getGlobalOpts(this);
        output(await createClient(opts).slack.sendMessage(o.connectionId, o), {
          json: !!opts.json,
        });
      }),
    );
  const actions = slack
    .command("action")
    .description("Inspect durable send status; sent is not delivered or read");
  connection(actions.command("get <action-id>")).action(
    withErrorHandler(async function (
      this: Command,
      id: string,
      o: { connectionId: string },
    ) {
      const opts = getGlobalOpts(this);
      output(await createClient(opts).slack.getAction(o.connectionId, id), {
        json: !!opts.json,
      });
    }),
  );
  const files = slack
    .command("file")
    .description("Read authorized file metadata and download bytes");
  connection(files.command("get <file-id>")).action(
    withErrorHandler(async function (
      this: Command,
      id: string,
      o: { connectionId: string },
    ) {
      const opts = getGlobalOpts(this);
      output(await createClient(opts).slack.getFile(o.connectionId, id), {
        json: !!opts.json,
      });
    }),
  );
  connection(files.command("download <file-id>"))
    .requiredOption(
      "--output <path>",
      "New output file path (will not overwrite)",
    )
    .action(
      withErrorHandler(async function (
        this: Command,
        id: string,
        o: { connectionId: string; output: string },
      ) {
        const opts = getGlobalOpts(this);
        const bytes = await createClient(opts).slack.downloadFile(
          o.connectionId,
          id,
        );
        await writeFile(o.output, bytes, { flag: "wx", mode: 0o600 });
        output({ path: o.output, bytes: bytes.length }, { json: !!opts.json });
      }),
    );
  registerSlackOperationCommands(slack, { conversations, messages, files });
}
